### PR TITLE
Rework breakdown condition in BiCGSTAB

### PR DESCRIPTION
**Pull request by Rombur**
This PR does two things: 
1. it reworks the breakdown conditions. Right we have an input parameter that is not used. Instead of the current condition, I've the same conditions than the ones used in Trilinos and in matlab.
2. in case of breakdown, the computation just stops. Neither Trilinos or matlab tries to keep going once there is a breakdown. I don't think there is anything we can do in case of breakdown, since it means that we divided by "zero" and thus, the computation doesn't make sense anymore.

closes https://github.com/dealii/dealii/issues/11338

### CODE REVIEW

**Comment by Rombur** [comment]
I've reverted the new breakdown conditions because some of our tests fail otherwise. Now we are using the old version but we use the tolerance given by the user.

**Review by kronbichler — APPROVED** [review]
Looks good to me.

File: include/deal.II/lac/solver_bicgstab.h [hunk 1]
```diff
@@ -418,9 +416,7 @@ SolverBicgstab<VectorType>::iterate(const MatrixType &        A,
 
       alpha = rho / rhobar;
 
-      // TODO:[?] Find better breakdown criterion
-
-      if (std::fabs(alpha) > 1.e10)
+      if (std::fabs(alpha) * additional_data.breakdown > 1)
```
Comment (bangerth) [inline include/deal.II/lac/solver_bicgstab.h:419]: Interesting. I *think* that must have been the intent all along.

**Review by bangerth — COMMENTED** [review]
I'm no expert in bicgstab, but my understanding is that breakdown is something that can actually happen and can be recovered from. I'm not sure, but I think https://mathworld.wolfram.com/BiconjugateGradientStabilizedMethod.html hints at that.

My understanding that breakdown in bicgstab is when our Krylov sequence has sort of run into a dead end, but that we can then re-start from there. I *think* that that's what the current implementation does, but I found it difficult to read through the class as it's written right now with the externalized `start()` and `iterate` functions -- maybe it would be simpler to inline these into the main function right away. 

Regardless, what I *think* is also currently happening is that we don't properly check for the number of iterations done and terminate after as many iterations as we have done, with each breakdown counting for one iteration.

**Comment by bangerth** [comment]
Take a look here for the breakdown issue: http://www.math.utep.edu/Faculty/xzeng/2017spring_math5330/2017spring_math5330/MATH_5330_Computational_Methods_of_Linear_Algebra_files/ln07.pdf

**Comment by bangerth** [comment]
Maybe we should just implement the algorithm from that PDF...

**Comment by Rombur** [comment]
To be honest with you I have never heard of restarted bicstab

> I'm no expert in bicgstab, but my understanding is that breakdown is something that can actually happen and can be recovered from. I'm not sure, but I think mathworld.wolfram.com/BiconjugateGradientStabilizedMethod.html hints at that.

Where do you see the hint to the restarted method.

> My understanding that breakdown in bicgstab is when our Krylov sequence has sort of run into a dead end, but that we can then re-start from there. I think that that's what the current implementation does, but I found it difficult to read through the class as it's written right now with the externalized start() and iterate functions -- maybe it would be simpler to inline these into the main function right away.

I am not sure how much we restart the method compared to simply "keep going" which you can still do by putting the breakdown tolerance to zero.

> Regardless, what I think is also currently happening is that we don't properly check for the number of iterations done and terminate after as many iterations as we have done, with each breakdown counting for one iteration.

Right we check the number of iterations inside `iterate()` but each restart count as one external iteration which doesn't matter since we never check the number of external iterations.

> Maybe we should just implement the algorithm from that PDF...

The condition in the PDF is the same that the one used by Trilinos and Matlab. Several tests and step 51 don't pass with this condition and no restart. Basically these implementation check that denominator of the fraction is not too small while we check that the ratio is not too large. That matters if the numerator is also small.

Here is what I propose, we either keep the implementation in this PR and if you want a "restart" you put the breakdown tolerance so the algorithm keeps going until you reach the maximum number of iterations or I implement the same breakdown conditions than Trilinos and a proper restarting strategy. What do you think?

**Comment by bangerth** [comment]
Let's do it right. I think that the implementation we have currently is fundamentally broken. Let's implement something like what Trilinos does + restart. Does that seem reasonable?

**Comment by Rombur** [comment]
Sounds good.

**Comment by Rombur** [comment]
This ready for review. Basically, our algorithm is already the same as Belos. The difference is the breakdown conditions. What I didn't realize is how small the breakdown condition is in Trilinos. It's even worse in PETSc where the breakdown condition is that floating point is equal to zero!! I have fixed the restart condition and now the algorithm exits when we reached `max_iter`.

**Comment by Rombur** [comment]
The OSX failure has nothing to do with this PR. The build does not pass configuration because somehow the build system does not recognize that `AppleClang` support C++14

**Comment by masterleinad** [comment]
> The OSX failure has nothing to do with this PR. The build does not pass configuration because somehow the build system does not recognize that `AppleClang` support C++14

That sounds like https://github.com/dealii/dealii/pull/11305 and should be fixed by rebasing.

**Comment by Rombur** [comment]
@bangerth ping

**Review by bangerth — APPROVED** [review]

**Pull request merged**